### ISSUE TITLE
Cannot open the NuGet Package manager

### ISSUE THREAD

**Issue by vsfeedback** [NuGet/Home#10336]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/Cannot-open-Nuget-browser/1265371)._

---
[severity:I'm unable to use this version] [regression] [worked-in:16.8.2]
Managing Nuget packages > Browse brings up the following error
! [image.png] (https://aka.ms/dc/image?name=Bca4ae68667b4419392c520bf946e165d637416996804221774_20201123-144121-image.png&tid=ca4ae68667b4419392c520bf946e165d637416996804221774)

Have restarted computer/VS

VS: 16.8.2 
! [image.png] (https://aka.ms/dc/image?name=B7bd6f85befdd40a6aa964320715f6281637416997462030978_20201123-144227-image.png&tid=7bd6f85befdd40a6aa964320715f6281637416997462030978)

I did update from VS16.8.1 the other day, not sure if it was part of the upgrade or not

Any help would be appreciated

Thanks

---
#### Feedback Bot on 11/22/2020, 10:51 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by nkolev92** [comment]
I don't know if this is a 100% repro regression, given that there are no NuGet changes among 16.8.1 and 16.8.2, but the error message is clear. 

The issue is happening because of the codespaces changes. 

cc @rrelyea @dtivel 

cc @aortiz-msft This is a P1 on dev community...worth prioritizing an investigation I think.

**Comment by nkolev92** [comment]
https://developercommunity2.visualstudio.com/t/Internal-Error-with-NuGet-Package-Manage/1271505 is another issue with the same root cause.

**Comment by dtivel** [comment]
The NuGet version is 5.8.0.6930 (https://github.com/NuGet/NuGet.Client/commit/830c8be45dbbccd411ecf6080abff0c2c98079cf).

Here is the callstack:
```
StreamJsonRpc.RemoteInvocationException: An internal error occurred. Please contact Microsoft Support.
 at Microsoft.ServiceHub.Framework.ServiceJsonRpcDescriptor.LocalProxyGeneration.<ReturnedValueTaskOfTHelper>d__25`1.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at NuGet.VisualStudio.Internal.Contracts.ProjectContextInfo.<GetInstalledPackagesAsync>d__13.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at NuGet.PackageManagement.VisualStudio.PackageCollection.<FromProjectsAsync>d__6.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at NuGet.PackageManagement.UI.PackageItemLoader.<UpdateStateAndReportAsync>d__21.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at NuGet.PackageManagement.UI.InfiniteScrollList.<LoadNextPageAsync>d__55.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at NuGet.PackageManagement.UI.InfiniteScrollList.<LoadItemsCoreAsync>d__54.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at NuGet.PackageManagement.UI.InfiniteScrollList.<RepopulatePackageListAsync>d__50.MoveNext()
 RPC server exception: Microsoft.Assumes+InternalErrorException: An internal error occurred. Please contact Microsoft Support.
 at Microsoft.Assumes.Fail(String message)
 at NuGet.VisualStudio.Internal.Contracts.PackageReferenceContextInfo.Create(PackageReference packageReference)
 at NuGet.PackageManagement.VisualStudio.NuGetProjectManagerService.<>c.<GetInstalledPackagesAsync>b__10_2(PackageReference pr)
 at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
 at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
 at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
 at NuGet.PackageManagement.VisualStudio.NuGetProjectManagerService.<GetInstalledPackagesAsync>d__10.MoveNext()
 --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Threading.Tasks.ValueTask`1.get_Result()
 at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1.ConfiguredValueTaskAwaiter.GetResult()
 at Microsoft.ServiceHub.Framework.ServiceJsonRpcDescriptor.LocalProxyGeneration.<ReturnedValueTaskOfTHelper>d__25`1.MoveNext()
```
The fault is in [`PackageReferenceContextInfo.Create(PackageReference)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/PackageReferenceContextInfo.cs#L33) because `packageReference` is `null`.

I'm surprised that `NuGetProject.GetInstalledPackagesAsync(...)` can return null values.  Throughout NuGet code there is the assumption that the return value will be either an empty enumerable or a non-empty enumerable containing non-`null` values.  There are a few exceptions though like in [`MultiSourcePackageMetadataProvider.GetLatestPackageMetadataAsync(...)`](https://github.com/NuGet/NuGet.Client/pull/501/files#diff-8d181a5e8f30440442edd95cdf5a83c9fe1b27c9b02269a71358a6e48f2b8091R103).

**Comment by dtivel** [comment]
There are 8 product implementations of [`NuGetProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Core/NuGet.PackageManagement/Projects/NuGetProject.cs#L56):

*  [`FolderNuGetProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Core/NuGet.PackageManagement/Projects/FolderNuGetProject.cs#L90-L93)
*  [`InstallCommandProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Clients/NuGet.CommandLine/Common/InstallCommandProject.cs#L75-L85)
*  [`LegacyPackageReferenceProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs#L174-L179)
*  [`MSBuildNuGetProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Core/NuGet.PackageManagement/Projects/MSBuildNuGetProject.cs#L113-L116)
*  [`NetCorePackageReferenceProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/NetCorePackageReferenceProject.cs#L221-L231)
*  [`PackagesConfigNuGetProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Core/NuGet.PackageManagement/Projects/PackagesConfigNuGetProject.cs#L237-L240)
*  [`ProjectJsonNuGetProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L137-L157)
*  [`ProjectKNuGetProject.GetInstalledPackagesAsync(...)`](https://github.com/NuGet/NuGet.Client/blob/830c8be45dbbccd411ecf6080abff0c2c98079cf/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectKNuGetProject.cs#L182-L217)

`InstallCommandProject.GetInstalledPackagesAsync(...)` is only used by NuGet.exe not Visual Studio (PM UI); however, none of the implementations can return an enumerable containing `null`.

**Comment by kartheekp-ms** [comment]
https://developercommunity.visualstudio.com/content/problem/1284267/nuget-package-manager-broken.html and https://developercommunity.visualstudio.com/content/problem/1284275/manage-nuget-packages-for-solution-not-displaying.html issues are reported with the same root cause.

**Comment by donnie-msft** [comment]
Duplicate https://developercommunity2.visualstudio.com/t/Error-occurred-when-trying-to-Manage-N/1288583

**Comment by ehasis** [comment]
This is happening in one of my solutions. That one use project.json for nuget reference instead of PackageReference or package.config.

**Comment by zivkan** [comment]
@ehasis what version of Visual Studio are you using, and what project type are you using project.json with?  I don't know of any project system in Visual Studio 2019 that uses project.json.

**Comment by ehasis** [comment]
@zivkan I'm using 16.8.3, the project type is a huge VB.NET WinForms application, but is easy to simulate with a small one. We use project.json because we have to keep compatibility with VS 2015 since couple of our colleagues still use it.

**Comment by zivkan** [comment]
@ehasis after installing VS2015, finding the .NET Core 1.0 preview tooling and installing that, creating a sample solution with project.json and xproj, when I try to open the solution in Visual Studio 2019, I get an error saying the project is unsupported and `UpgradeLog.htm` says:

> Xproj project files are no longer supported. For details on migrating to csproj see https://docs.microsoft.com/en-us/dotnet/core/migration/.

I'm not sure you're in a supported scenario. Furthermore, .NET Core didn't support WinForms until .NET Core 3.0, but the VS2015 preview tools for .NET Core only supports .NET Core 1.0. I don't understand the solution/projects you're describing. Are you able to create a new sample solution and attach it here?

**Comment by ehasis** [comment]
@zivkan easy to reproduce, here are the steps:
1. Open Visual Studio and create a new Console App (.NET Framework) (C#, I choose sdk 4.7.2)
2. Add a new file called project.json and paste the following content:
```json
{
  "dependencies": {
    "Newtonsoft.Json": "12.0.3"
  },
  "frameworks": {
    "net472": {}
  },
  "runtimes": {
    "win": {}
  }
}
```
3. Save All, compile and close the Visual Studio
4. Open the Visual Studio again and open the solution
5. Right click on the solution and choose Manage NuGet Package for the solution
6. Click on Browse tab and the error will show

**Comment by zivkan** [comment]
@dtivel  @rrelyea @sbanni, using the above instructions, I tracked down the issue to this assumption being incorrect about project.json projects: https://github.com/NuGet/NuGet.Client/blob/b52bb6fe102c2a362c8adc6965cbb4c192781b1b/src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ContextInfos/PackageReferenceContextInfo.cs#L18

@ehasis, I recommend you consider moving to packages.config if you want to keep compatibility with VS2015, or move to PackageReference and drop support for VS2015 if that's more important. Project.json usage is low (0.2% of project in VS2019), so I expect we're going to drop support for it in the future. I guess in the next major version, but that decision is above my paygrade. I've marked this issue as a regression, and we're generally not allowed to regress features during the same major version of VS, so I expect this to be fixed. But setting expectations, this feature is low usage and if it causes too much technical debt, people are going to need to make tough decisions.

**Comment by dtivel** [comment]
@ehasis, thank you very much for the repro steps!  I have a fix in code review.

**Comment by pdario66** [comment]
@dtivel I have just upgraded to version 16.8.4, but the problem is still sitting there, when will be this released?

![VS 16_8_4 problems with NuGet Manager](https://user-images.githubusercontent.com/77633072/104941265-2170f100-59b3-11eb-85fe-3f6f5883656d.png)

**Comment by baepiff** [comment]
@dtivel I upgraded to 16.8.5. But the issue is still there. Could you share the release date?
![image](https://user-images.githubusercontent.com/1399501/108648201-fb80c400-74fd-11eb-81f6-84a9b9242b0b.png)

**Comment by pdario66** [comment]
@dtivel - I do not know exactly where to get the release date, but I copied all the info (that also contain the build number)

Microsoft Visual Studio Professional 2019
Version 16.8.5
VisualStudio.16.Release/16.8.5+31005.135
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Visual C++ 2019   00435-60000-00000-AA199
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.8.559.8768
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2052803+84e121f1403378489b842e1797df2f3f5a49ac3c
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.8.559.8768
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.8.559.8768
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.8.559.8768
Azure Functions and Web Jobs Tools

C# Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

JetBrains ReSharper Ultimate 2020.1.4   Build 201.0.20200702.125931
JetBrains ReSharper Ultimate package for Microsoft Visual Studio. For more information about ReSharper Ultimate, visit http://www.jetbrains.com/resharper. Copyright © 2021 JetBrains, Inc.

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30924.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.8.1
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62102.01130
Microsoft SQL Server Data Tools

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.21016.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.8.0-beta.20507.4+da6be68280c89131cdba2045525b80890401defd
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

**Comment by kartheekp-ms** [comment]
@baepiff @pdario66 

This fix is currently in `16.9` preview versions and can be downloaded from here https://visualstudio.microsoft.com/vs/preview/.

**Comment by pdario66** [comment]
@kartheekp-ms @baepiff 
Thanks, I can wait and resist until 16.9 is released :) 
Regards

**Comment by herb8work** [comment]
How can I get that to work with packages.config in VS 16.8.6 without installing a prerelese version of VS?

**Comment by ehasis** [comment]
@herb8work if you use project.json, you can add the dependency directly editing the file by hand. Search the package name and version on the nuget.org in the web browser.

**Comment by pdario66** [comment]
@ehasis yes we all know about that - but we have a solution with different projects over than 100 and they are mixed types c# and C++ projects some uses json and some else uses packages.config - what it really bothers here is that at the end you need to consolidate the solution and with this  broken is painful

**Comment by ehasis** [comment]
@pdario66 I understand, we have the same situation here. If you need edit all at once, as a temporary workaround until the solution is in the GA, you could try install the Preview in one machine and do this kind of maintenance there.

**Comment by nkolev92** [comment]
Hey all, 

Have you considered moving your project.json projects to PackageReference?
Any blockers for that?

I know it's not an ideal workaround, but it puts you in an actively developed world, rather than project.json which is something that never RTM'ed on its own.

**Comment by herbutant** [comment]
> Have you considered moving your project.json projects to PackageReference?

I changed that in NuGet settings UI in VS. But that had no no notable effect.

**Comment by zivkan** [comment]
That setting only affects when a non-SDK style .NET project installs the first package. Once a project contains at least one package, the project keeps using that package style. If you uninstall all packages in a project, unfortunately you'll also need to reload the project to "reset" NuGet's memory of the project.

Some projects that don't support PackageReference will also ignore the setting, and always use packages.config. SDK style projects only support PackageReference, hence will also ignore the setting in case packages.config is selected.

**Comment by pdario66** [comment]
> 
> 
> @baepiff @pdario66
> 
> This fix is currently in `16.9` preview versions and can be downloaded from here https://visualstudio.microsoft.com/vs/preview/.

@kartheekp-ms  - Thanks - today I installed the 16.9.2 and it is all working ;)
